automation/workflow/store.go

```diff
@@ -0,0 +1,304 @@
+	Undeploy(id ID) error
+}
+
+// A Lister lists entries from a workflow store
+type Lister interface {
+	List(ctx context.Context, limit, offset int) ([]*Workflow, error)
+	ListDeployed(ctx context.Context, limit, offset int) ([]*Workflow, error)
+}
+
+// MemStore is an in memory representation of a Store
+type MemStore struct {
+	mu        *sync.Mutex
+	workflows map[ID]*Workflow
+}
+
+var _ Store = (*MemStore)(nil)
+
+// NewMemStore return a MemStore
+func NewMemStore() *MemStore {
+	return &MemStore{
+		mu:        &sync.Mutex{},
+		workflows: map[ID]*Workflow{},
+	}
+}
+
+// Create creates a new Workflow and adds it to the Store. It does not check that
+// the given dataset or peer ids are valid, beyond that they are not empty
+// There should only be one workflow per dataset id
+func (m *MemStore) Create(did string, pid profile.ID, triggers []Trigger) (*Workflow, error) {
+	if did == "" {
+		return nil, fmt.Errorf("dataset ID required")
+	}
+	if pid == "" {
+		return nil, fmt.Errorf("profile ID required")
+	}
+	if _, err := m.GetDatasetWorkflow(did); !errors.Is(err, ErrNotFound) {
+		return nil, ErrWorkflowForDatasetExists
+	}
+
+	now := time.Now()
+	wf := &Workflow{
+		ID:        NewID(),
+		DatasetID: did,
+		OwnerID:   pid,
+		Created:   &now,
+		Triggers:  triggers,
+	}
+	m.mu.Lock()
+	m.workflows[wf.ID] = wf
+	m.mu.Unlock()
+	return wf, nil
+}
+
+// Get fetches a Workflow using the associated ID
+func (m *MemStore) Get(wid ID) (*Workflow, error) {
+	m.mu.Lock()
+	wf, ok := m.workflows[wid]
+	m.mu.Unlock()
+	if !ok {
+		return nil, ErrNotFound
+	}
+	return wf, nil
+}
+
+// GetDatasetWorkflow fetches a workflow using the dataset ID
+func (m *MemStore) GetDatasetWorkflow(did string) (*Workflow, error) {
+	if did == "" {
+		return nil, ErrNotFound
+	}
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	for _, wf := range m.workflows {
+		if wf.DatasetID == did {
+			return wf, nil
+		}
+	}
+	return nil, ErrNotFound
+}
+
+// Update updates a workflow with changes
+func (m *MemStore) Update(wf *Workflow) error {
+	if wf == nil {
+		return ErrNilWorkflow
+	}
+	if _, err := m.Get(wf.ID); errors.Is(err, ErrNotFound) {
+		return ErrNotFound
+	}
+	m.mu.Lock()
+	m.workflows[wf.ID] = wf
+	m.mu.Unlock()
+	return nil
+}
+
+// Remove removes a workflow from a store
+func (m *MemStore) Remove(id ID) error {
+	m.mu.Lock()
+	_, ok := m.workflows[id]
+	if !ok {
+		return ErrNotFound
+	}
+	delete(m.workflows, id)
+	m.mu.Unlock()
+	return nil
+}
+
+// Deploy enables the workflow to be run by an orchestrator
+func (m *MemStore) Deploy(id ID) error {
+	wf, err := m.Get(id)
+	if err != nil {
+		return err
+	}
+	wf.Deployed = true
+	return m.Update(wf)
+}
+
+// Undeploy prevents the workflow from being run by an orchestrator
+func (m *MemStore) Undeploy(id ID) error {
+	wf, err := m.Get(id)
+	if err != nil {
+		return err
+	}
+	wf.Deployed = false
+	return m.Update(wf)
+}
+
+// List lists all the workflows in the store, by decending order from time of
+// creation
```

Comment:
Want to expand the description in the interface definition for `List` and `ListDeployed` to include expectations for limit and offset. These are the assumptions I'm going off of, but would like feedback:
- limit = -1 && offset = 0 means `fetchAll`
- negative limit or negative offset returns an error

Main question is around the offset being greater than the size of the list. Should that be an error or an empty list? I have an error coded up here, but looking at it an empty list is probably the way to go.

Reply:
totally agree on `limit: -1, offset: 0` as the fetchAll sentinel value. I'm working on a `params.ListAll` that does exactly the same thing: https://github.com/qri-io/qri/blob/3133469fe112b8590ee8ddb5a9ba19f2adbe9c11/params/list.go#L4-L6

very much agree that offsets greater than the length of the list should return an empty slice & no error